lib/rdoc/markup/heading.rb

```diff
@@ -1,84 +1,100 @@
 # frozen_string_literal: true
-##
-# A heading with a level (1-6) and text
 
-RDoc::Markup::Heading =
-  Struct.new :level, :text do
-
-  @to_html = nil
-  @to_label = nil
-
-  ##
-  # A singleton RDoc::Markup::ToLabel formatter for headings.
-
-  def self.to_label
-    @to_label ||= RDoc::Markup::ToLabel.new
-  end
-
-  ##
-  # A singleton plain HTML formatter for headings.  Used for creating labels
-  # for the Table of Contents
-
-  def self.to_html
-    return @to_html if @to_html
-
-    markup = RDoc::Markup.new
-    markup.add_regexp_handling RDoc::CrossReference::CROSSREF_REGEXP, :CROSSREF
-
-    @to_html = RDoc::Markup::ToHtml.new nil
-
-    def @to_html.handle_regexp_CROSSREF(target)
-      target.text.sub(/^\\/, '')
+module RDoc
+  class Markup
+    # A heading with a level (1-6) and text
+    #
+    #  RDoc syntax:
+    #   = Heading 1
+    #   == Heading 2
+    #   === Heading 3
+    #
+    #  Markdown syntax:
+    #   # Heading 1
+    #   ## Heading 2
+    #   ### Heading 3
+    class Heading < Element
+      #: String
+      attr_reader :text
+
+      #: Integer
+      attr_accessor :level
+
+      # A singleton RDoc::Markup::ToLabel formatter for headings.
+      #: () -> RDoc::Markup::ToLabel
+      def self.to_label
+        @to_label ||= Markup::ToLabel.new
+      end
+
+      # A singleton plain HTML formatter for headings. Used for creating labels for the Table of Contents
+      #: () -> RDoc::Markup::ToHtml
+      def self.to_html
+        @to_html ||= begin
+          markup = Markup.new
+          markup.add_regexp_handling CrossReference::CROSSREF_REGEXP, :CROSSREF
+
+          @to_html = Markup::ToHtml.new nil
+
+          def @to_html.handle_regexp_CROSSREF(target)
+            target.text.sub(/^\\/, '')
+          end
+
+          @to_html
+        end
+      end
+
+      #: (Integer, String) -> void
+      def initialize(level, text)
+        @level = level
+        @text = text
+        super()
```

Comment:
Initializing the base object (`Markup::Element`) before this object may be better:

```suggestion
        super()
        @level = level
        @text = text
```

Or we may want to remove `super()` here because the current `Markup::Element#initialize` is empty.

Reply:
I moved the `super()` call above. I would keep it as if we end up adding shared state to the parent class, then we don't need to update all of the children.